modules/tinymce/src/plugins/directionality/main/ts/core/Direction.ts

```diff
@@ -5,28 +5,43 @@
  * For commercial licenses see https://www.tiny.cloud/
  */
 
-import Editor from 'tinymce/core/api/Editor';
-import Tools from 'tinymce/core/api/util/Tools';
+import { Arr } from '@ephox/katamari';
+import { Traverse, Attribute, SugarElement } from '@ephox/sugar';
 
-const setDir = (editor: Editor, dir: string) => {
-  const dom = editor.dom;
-  let curDir;
-  const blocks = editor.selection.getSelectedBlocks();
+import Editor from 'tinymce/core/api/Editor';
 
-  if (blocks.length) {
-    curDir = dom.getAttrib(blocks[0], 'dir');
+type Dir = 'rtl' | 'ltl';
 
-    Tools.each(blocks, (block) => {
-      // Add dir to block if the parent block doesn't already have that dir
-      if (!dom.getParent(block.parentNode, '*[dir="' + dir + '"]', dom.getRoot())) {
-        dom.setAttrib(block, 'dir', curDir !== dir ? dir : null);
+const setDir = (editor: Editor, dir: Dir) => {
+  const selectedBlocks = editor.selection.getSelectedBlocks();
+  Arr.each(selectedBlocks, (block) => {
+    const sugarBlock = SugarElement.fromDom(block);
+    const blockParent = Traverse.parent(sugarBlock);
+    blockParent.map((blockParent: SugarElement<Element>) => {
+      const blockParentDirection = Attribute.get(blockParent, 'dir');
```

Comment:
This is a regression from the previous behaviour (though the previous behaviour also had a bug). The `editor.dom.getParent` function will get any ancestor that matches the specified selector. As such, this needed to check multiple parent elements, not just the immediate parent. So the equivalent to the old logic would have been `SelectorFind.closest(sugarBlock, '*[dir="' + dir + '"]', (elem) => Compare.eq(elem, SugarElement.fromDom(editor.dom.getRoot()))`.

We can do this more easily though, as we can use computed CSS values to get the directionality. As such, we can fix the regression by just changing this to the following:
```suggestion
      const blockParentDirection = Direction.get(blockParent);
```
This uses the Sugar `Direction` module (which isn't currently imported) to determine the current computed directionality.

Reply:
Another bug I found while reviewing is this doesn't take into account where we need to apply the directionality. In the `li` case, the parent here will be the `ul` or `ol`. So checking the current directionality on that would be incorrect and instead we need to look at the `ul` or `ol` parent instead since that's where we apply the `dir` attribute.

Reply:
I fixed that with the last commit 👍